myelin/kernel/cuda-arithmetic.cc

```diff
@@ -0,0 +1,618 @@
+    const Shape &shape = step->output(0)->shape();
+    for (auto *input : step->inputs()) {
+      if (input->type() != type) return false;
+      if (!input->Compatible(step->output(0))) return false;
+
+      // NB: general broadcasting not yet supported.
+      if (input->elements() != shape.elements() && input->elements() != 1) {
+        return false;
+      }
+    }
+    for (auto *output : step->outputs()) {
+      if (output->type() != type) return false;
+      if (output->shape() != shape) return false;
+      if (output->elements() != shape.elements()) return false;
+    }
+
+    // Check that element type is supported by CUDA.
+    if (TypeTraits::of(type).ptx() == nullptr) return false;
+
+    // Dense encoding required.
+    for (auto *input : step->inputs()) input->RequireDense();
+    for (auto *output : step->outputs()) output->RequireDense();
+
+    return true;
+  }
+
+  void Adjust(Step *step) override {
+    // Inputs and ouputs must be in standard format.
+   for (auto *input : step->inputs()) {
+      input->RequireDense();
+      input->RequireStandardOrder();
+    }
+    for (auto *output : step->outputs()) {
+      output->RequireDense();
+      output->RequireStandardOrder();
+    }
+
+    // Enable sharing of inputs and outputs.
+    for (int i = 0; i < step->indegree(); ++i) {
+      for (int j = 0; j < step->outdegree(); ++j) {
+        if (step->input(i)->shape() == step->output(j)->shape()) {
+          if (step->AllowInPlace(i, j)) break;
+        }
+      }
+    }
+  }
+
+  void GeneratePTX(Step *step, PTXMacroAssembler *ptx)  override {
+    // Parse expression for evaluation.
+    Express expr(Express::NVIDIA);
+    InitExpression(step, &expr, true);
+
+    // Set grid size. Use one thread for each element.
+    Tensor *output = step->output(0);
+    int size = output->elements();
+    ptx->set_grid_dims(size);
+
+    // Set up compilation state.
+    Compilation comp;
+    comp.step = step;
+    comp.ptx = ptx;
+
+    // Get element type.
+    comp.dtype = output->type();
+    const TypeTraits &traits = TypeTraits::of(comp.dtype);
+    comp.type = traits.ptx();
+    comp.size = traits.size();
+
+    // Optimize expression.
+    expr.EliminateCommonSubexpressions();
+    expr.CacheResults();
+
+    // Rewrite expression.
+    Express instrs;
+    CHECK(expr.Rewrite(ptx_model, &instrs));
+    instrs.ComputeLiveRanges();
+    int regs = instrs.AllocateRegisters();
+
+    // Get grid location.
+    ptx_decl(b32, idx);
+    ptx->GetThreadIndex(idx, 0);
+
+    // Check bounds.
+    ptx_decl(pred, outside);
+    ptx_emit(setp.ge.u32, outside, idx, PTXImm(size));
+    ptx_if(outside);
+    ptx_jump(done);
+    ptx_endif();
+
+    // Compute element offset.
+    ptx_decl(b64, offset);
+    comp.offset = offset;
+    ptx_emit(mul.wide.u32, offset, idx, PTXImm(comp.size));
+    ptx_decl(b64, addr);
+    comp.addr = addr;
+
+    // Allocate registers.
+    comp.reg.resize(regs);
+    std::vector<PTXReg> reg(regs);
```

Comment:
This vector doesn't seem to be used.

Reply:
You're right. The regs are now in the comp block. Removed.